File: 20-actors/watari/methods/autorun.py
#!/usr/bin/env python3
"""autorun.py — watari AUTONOMOUS live-moving-craft situational-awareness heartbeat on the kotoba
Datom log. ADR-2606041827.

This is what "kotoba で自律的に稼働する" means for watari, in the constitution-permitted form
(mirrors shionome / ipaddress / yabai / sukashi / watatsuna autorun). Each heartbeat the actor runs
its whole SITUATIONAL-AWARENESS pipeline ITSELF, with no human in the loop:

    observe (load the OFFLINE merged craft graph) → classify
      → analyze (latest as-of fix per craft → lane/corridor load by kind → chokepoint transit →
        approach congestion → freshness tail — aggregate-first, G2 awareness not surveillance)
      → PERSIST a content-addressed transaction to the append-only kotoba Datom log
        (graph datoms + derived :movement/* signals), linking the previous tx's CID.

Constitutional posture holds by construction: outputs are aggregate lane/chokepoint/approach
density framed toward safety + congestion-easing — NEVER a "follow this craft" / targeting feed
(G2); a craft is a craft, NEVER a person — no track is linked to a named individual, no
pattern-of-life (G4, the defining gate). The chokepoint-transit output composes with watatsuna's
static cable-load over the SAME chokepoint keywords (静↔動 maritime resilience).

The loop is deterministic / resume-safe (cycle drives tx-id + as-of → same CIDs) and append-only
(the fix stream IS the trajectory; 非終末論). WHAT STAYS GATED (G7): it NEVER pulls a live
AISStream / OpenSky / adsb.fi feed and NEVER pushes to a live kotoba node. Ingest is the offline
merged/seed graph; persistence is the LOCAL append-only log. Live ingest is Council + operator
gated. Stdlib only.
"""
from __future__ import annotations

import argparse
import os
import pathlib
import sys

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from analyze import analyze, classify, load_edn  # noqa: E402
from kotoba import (append_tx, derived_datoms, graph_datoms, head_cid,  # noqa: E402
                    make_tx, read_log, verify_chain)

HERE = pathlib.Path(__file__).resolve().parent
DATA = HERE.parent / "data"
MERGED = DATA / "craft-graph.merged.kotoba.edn"
SEED = DATA / "seed-craft-graph.kotoba.edn"
LOG = DATA / "watari.datoms.kotoba.edn"
BASE_AS_OF = 20260608


def _graph_path(graph_path: pathlib.Path | None) -> pathlib.Path:
    if graph_path is not None:
        return graph_path
    return MERGED if MERGED.exists() else SEED


def run_cycle(cycle: int, graph_path: pathlib.Path | None = None,
              log_path: pathlib.Path = LOG) -> dict:
    """One autonomous heartbeat: observe → classify → analyze → persist a content-addressed Datom
    transaction (graph + derived :movement/* signals). cycle drives tx-id + as-of."""
    rows = load_edn(_graph_path(graph_path))      # observe — OFFLINE merged graph (G7: no live feed)
    craft, fixes, legs, lanes = classify(rows)
    a = analyze(craft, fixes, legs, lanes)        # aggregate situational-awareness signal (G2)
    datoms = graph_datoms(rows) + derived_datoms(craft, lanes, a)
    tx = make_tx(datoms, tx_id=cycle, as_of=BASE_AS_OF + cycle, prev_cid=head_cid(log_path))
    cid = append_tx(tx, log_path)                 # PERSIST to append-only LOCAL kotoba log
    top_choke = max(a["choke_transit"], key=lambda k: a["choke_transit"][k]) if a["choke_transit"] else "—"
    return {
        "cycle": cycle,
        "craft": len(craft),
        "fixes": len(fixes),
        "lanes": len(lanes),
        "chokepoints": len(a["choke_transit"]),
        "top_chokepoint": top_choke,
        "stale": len(a["stale"]),
        "datoms": len(datoms),
        "cid": cid,
    }


def run_autonomous(cycles: int = 3, graph_path: pathlib.Path | None = None,
                   log_path: pathlib.Path = LOG) -> dict:
    beats = [run_cycle(c, graph_path, log_path) for c in range(1, cycles + 1)]
    return {
        "cycles": cycles,
        "beats": beats,
        "log_length": len(read_log(log_path)),
        "head_cid": head_cid(log_path),
        "chain": verify_chain(log_path),
    }


if __name__ == "__main__":
    ap = argparse.ArgumentParser(description="watari autonomous moving-craft situational-awareness loop")
    ap.add_argument("--cycles", type=int, default=3, help="number of self-paced heartbeats")
    ap.add_argument("--graph", type=pathlib.Path, default=None, help="craft graph EDN (offline)")
    ap.add_argument("--log", type=pathlib.Path, default=LOG, help="kotoba Datom log path")
    ap.add_argument("--fresh", action="store_true", help="start a fresh log (remove existing)")
    args = ap.parse_args()
    if args.fresh and args.log.exists():
        args.log.unlink()
    res = run_autonomous(args.cycles, graph_path=args.graph, log_path=args.log)
    print("# watari — AUTONOMOUS moving-craft situational-awareness over the kotoba Datom log "
          "(offline ingest, LOCAL persist; live AIS/ADS-B feed / live-node push stays G7-gated)\n")
    for bt in res["beats"]:
        print(f"  ♥ cycle {bt['cycle']}: {bt['craft']} craft / {bt['fixes']} fixes / {bt['lanes']} lanes "
              f"· chokepoints {bt['chokepoints']} (top {bt['top_chokepoint']}) · stale-tail {bt['stale']} "
              f"+{bt['datoms']} datoms → cid {bt['cid'][:14]}…")
    ch = res["chain"]
    print(f"\n  log: {res['log_length']} tx · head {res['head_cid'][:14]}… · "
          f"chain {'OK ✓' if ch['ok'] else 'BROKEN at ' + str(ch['broken_at'])} · "
          f"situational-awareness, no person-tracking (G2/G4)")

File: 20-actors/watari/methods/autorun.clj
#!/usr/bin/env bb
;; Working Clojure port of methods/autorun.py — the autonomous moving-craft heartbeat.
(ns watari.methods.autorun
  "autorun.clj — watari AUTONOMOUS moving-craft situational-awareness heartbeat on the kotoba
  Datom log. ADR-2606041827.

  Each heartbeat: observe (OFFLINE merged craft graph, G7 no live feed) → classify → analyze
  (aggregate lane/chokepoint situational signal, G2) → PERSIST a content-addressed transaction
  (graph datoms + derived :movement/*) to the append-only kotoba Datom log, linking the previous
  CID. Deterministic / resume-safe; no external I/O; live AIS/ADS-B feed + live-node push stay
  G7-gated. A craft is a craft, never a person (G4).

  Run:  bb --classpath 20-actors 20-actors/watari/methods/autorun.clj --cycles 3 --fresh"
  (:require [watari.methods.analyze :as a]
            [watari.methods.kotoba :as k]
            [clojure.java.io :as io]
            [clojure.string :as str]))

(def ^:private this-file *file*)
(defn- data-dir [] (-> this-file io/file .getAbsoluteFile .getParentFile .getParentFile (io/file "data")))
(defn- merged [] (io/file (data-dir) "craft-graph.merged.kotoba.edn"))
(defn- seed [] (io/file (data-dir) "seed-craft-graph.kotoba.edn"))
(defn- default-log [] (io/file (data-dir) "watari.datoms.kotoba.edn"))
(def BASE-AS-OF 20260608)

(defn- graph-path [gp] (or gp (let [m (merged)] (if (.exists m) m (seed)))))

(defn run-cycle
  [cycle & {:keys [graph-path* log-path]}]
  (let [log-path (or log-path (default-log))
        rows (a/load-edn (graph-path graph-path*))
        g (a/classify rows)
        an (a/analyze g)
        datoms (vec (concat (k/graph-datoms rows) (k/derived-datoms g an)))
        tx (k/make-tx datoms :tx-id cycle :as-of (+ BASE-AS-OF cycle) :prev-cid (k/head-cid log-path))
        cid (k/append-tx tx log-path)
        top-choke (if (seq (:choke-transit an)) (key (apply max-key val (:choke-transit an))) "—")]
    {:cycle cycle :craft (count (:craft g)) :fixes (count (:fixes g)) :lanes (count (:lanes g))
     :chokepoints (count (:choke-transit an)) :top-chokepoint top-choke
     :stale (count (:stale an)) :datoms (count datoms) :cid cid}))

(defn run-autonomous [& {:keys [cycles graph-path* log-path] :or {cycles 3}}]
  (let [log-path (or log-path (default-log))
        beats (mapv #(run-cycle % :graph-path* graph-path* :log-path log-path) (range 1 (inc cycles)))]
    {:cycles cycles :beats beats :log-length (count (k/read-log log-path))
     :head-cid (k/head-cid log-path) :chain (k/verify-chain log-path)}))

(defn -main [& argv]
  (let [args (vec argv)
        cyc-idx (.indexOf args "--cycles")
        cycles (if (>= cyc-idx 0) (Integer/parseInt (nth args (inc cyc-idx))) 3)
        log-idx (.indexOf args "--log")
        log-path (if (>= log-idx 0) (io/file (nth args (inc log-idx))) (default-log))]
    (when (and (some #{"--fresh"} args) (.exists (io/file log-path)))
      (.delete (io/file log-path)))
    (let [res (run-autonomous :cycles cycles :log-path log-path)]
      (println (str "# watari — AUTONOMOUS moving-craft situational-awareness over the kotoba Datom log "
                    "(offline ingest, LOCAL persist; live AIS/ADS-B / live-node push stays G7-gated)\n"))
      (doseq [bt (:beats res)]
        (println (format "  ♥ cycle %d: %d craft / %d fixes / %d lanes · chokepoints %d (top %s) · stale %d +%d datoms → cid %s…"
                         (:cycle bt) (:craft bt) (:fixes bt) (:lanes bt) (:chokepoints bt)
                         (:top-chokepoint bt) (:stale bt) (:datoms bt) (subs (:cid bt) 0 14))))
      (let [ch (:chain res)]
        (println (format "\n  log: %d tx · head %s… · chain %s · situational-awareness, never surveillance (G2/G4)"
                         (:log-length res) (subs (:head-cid res) 0 14)
                         (if (:ok ch) "OK ✓" (str "BROKEN at " (:broken-at ch)))))))))

(when (= *file* (System/getProperty "babashka.file"))
  (apply -main *command-line-args*))
